src/lib/operators/projection.cpp

```diff
@@ -240,15 +180,47 @@ std::shared_ptr<const Table> Projection::_on_execute() {
     }
   }
 
-  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count_input_table; ++chunk_id) {
+  // Create the actual chunks, and, if needed, fill the projection_result_table. Also set MVCC and
+  // individually_sorted_by information as needed.
+  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
     const auto input_chunk = input_table.get_chunk(chunk_id);
     Assert(input_chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
 
+    auto projection_result_segments = Segments{};
+    const auto entire_chunk_pos_list = std::make_shared<EntireChunkPosList>(chunk_id, input_chunk->size());
+    for (auto column_id = ColumnID{0}; column_id < expressions.size(); ++column_id) {
+      // Turn newly generated ValueSegments into ReferenceSegments, if needed
+      if (expressions[column_id]->type != ExpressionType::PQPColumn && output_table_type == TableType::References) {
+        projection_result_segments.emplace_back(output_segments_by_chunk[chunk_id][column_id]);
+
+        const auto projection_result_column_id =
+            ColumnID{static_cast<ColumnID::base_type>(projection_result_segments.size() - 1)};
+
+        output_segments_by_chunk[chunk_id][column_id] = std::make_shared<ReferenceSegment>(
+            projection_result_table, projection_result_column_id, entire_chunk_pos_list);
+      }
+    }
+
     // The output chunk contains all rows that are in the stored chunk, including invalid rows. We forward this
     // information so that following operators (currently, the Validate operator) can use it for optimizations.
-    const auto chunk = std::make_shared<Chunk>(std::move(output_chunk_segments[chunk_id]), input_chunk->mvcc_data());
-    chunk->increase_invalid_row_count(input_chunk->invalid_row_count());
-    chunk->finalize();
+    auto chunk = std::shared_ptr<Chunk>{};
+    if (output_table_type == TableType::Data) {
+      chunk = std::make_shared<Chunk>(std::move(output_segments_by_chunk[chunk_id]), input_chunk->mvcc_data());
+      chunk->increase_invalid_row_count(input_chunk->invalid_row_count());
+      chunk->finalize();
```

Comment:
I don't know if that could become a performance issue anytime soon but `finalize` iterates over all of the `mvcc_data` to find `max_begin_cid`. In theory, we should be able to reuse `input_chunk`'s value, right? I don't know if it would be worth it though.

Reply:
See #2131 and #2137. The semantics of finalize are really unclear anyway.

Calculating `max_begin_cid` should not be an issue though as it is transferred as `_mvcc_data->max_begin_cid`:

```
if (has_mvcc_data() && !_mvcc_data->max_begin_cid) {
// calculate it
}
```

Reply:
Right, thanks!